File: src/documentPolicy.ts
/**
 * documentPolicy.ts — Document-Policy response-header audit.
 * T76 cycle 60.
 *
 * Document-Policy is a 2024-shipped W3C header (companion to
 * Permissions-Policy) that lets a document opt INTO additional
 * runtime constraints on its own features. Unlike Permissions-
 * Policy (which controls cross-origin access to powerful APIs
 * like camera, mic, geolocation), Document-Policy controls
 * features within the document itself:
 *
 *   * `document-write=?0` — disable document.write/writeln.
 *     Eliminates a classic DOM-XSS sink that also blocks
 *     async parsing. Modern apps shouldn't need it.
 *
 *   * `force-load-at-top=?1` — disable scroll restoration.
 *     Always load at the top of the page, never restore
 *     scroll position. Predictable UX for spa-like content.
 *
 *   * `unsized-media=?0` — require explicit width/height on
 *     <img> / <video> / <iframe>. Reduces CLS (Cumulative
 *     Layout Shift, a Core Web Vital).
 *
 *   * `oversized-images=?<ratio>` — reject images whose
 *     transmitted dimensions exceed the displayed
 *     dimensions by the given ratio. Catches "1080p hero
 *     image displayed at 200px" perf bugs.
 *
 *   * `js-profiling=?1` — opt INTO the JS Self-Profiling API.
 *     Lets the page collect performance.measureUserAgentSpecificMemory
 *     and JSCpuProfiler samples — observability primitive.
 *
 *   * `frame-loading=eager|lazy` — default iframe loading mode.
 *
 * Header reporting variant `Document-Policy-Report-Only:` is
 * also accepted by the spec for canary deploys.
 *
 * Findings:
 *
 *   - document-policy.missing                   warn
 *     No header. The document operates under default
 *     behaviour (document.write allowed, oversized images
 *     permitted, scroll restoration on, etc). For a
 *     supersociety baseline, set at minimum:
 *       Document-Policy: document-write=?0, force-load-at-top
 *
 *   - document-policy.permits-document-write    warn
 *     Header set but doesn't include `document-write=?0`.
 *     document.write is a parser-blocking DOM-XSS sink;
 *     modern apps should explicitly disable.
 *
 *   - document-policy.invalid                   warn
 *     Header value doesn't parse as Structured Fields
 *     (RFC 8941) dictionary. Browsers reject silently.
 *
 * Out of scope:
 *   * Localhost (consistent with the response-header detector
 *     family — local dev servers rarely set the full security
 *     header suite).
 *   * http pages — the header still works but the page has
 *     bigger problems.
 *
 * Reads from the same `topLevelResponseHeaders` Map as the
 * other response-header detectors. FOURTEENTH consumer of the
 * shared capture path. Uses the cycle-24
 * `responseHeaderDetector` helper.
 */

export interface DocumentPolicyFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface DocumentPolicySnapshot {
  pageUrl: string;
  pageIsLocalhost: boolean;
  /** Raw `Document-Policy` value (trimmed), or null. */
  raw: string | null;
  /**
   * Parsed directives. Each key is lowercase; value is the raw
   * structured-field token (e.g. '?0', '?1', '0.5'). Empty
   * map when header is null or unparseable.
   */
  directives: Record<string, string>;
}

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return (
      h === 'localhost' ||
      h === '127.0.0.1' ||
      h === '::1' ||
      h.endsWith('.localhost')
    );
  } catch {
    return false;
  }
}

function getHeader(
  headers: Record<string, string> | undefined,
  name: string,
): string | null {
  if (!headers) return null;
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === name) return v;
  }
  return null;
}

/**
 * Best-effort parse of a Document-Policy header value. The
 * Structured Fields spec defines this as a Dictionary; we
 * implement just enough to support the common forms:
 *   `document-write=?0, force-load-at-top, oversized-images=2.0`
 * is parsed as:
 *   { 'document-write': '?0', 'force-load-at-top': '?1',
 *     'oversized-images': '2.0' }
 *
 * Bare keys imply `?1` (RFC 8941 §3.1.2). Token values are
 * kept as raw strings; we don't validate ranges.
 */
function parseDirectives(raw: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const part of raw.split(',')) {
    const trimmed = part.trim();
    if (!trimmed) continue;
    const eq = trimmed.indexOf('=');
    if (eq === -1) {
      out[trimmed.toLowerCase()] = '?1';
    } else {
      const key = trimmed.slice(0, eq).trim().toLowerCase();
      const value = trimmed.slice(eq + 1).trim();
      if (key) out[key] = value;
    }
  }
  return out;
}

export function buildDocumentPolicySnapshot(
  pageUrl: string,
  headers: Record<string, string> | undefined,
): DocumentPolicySnapshot {
  const pageIsLocalhost = isLocalhost(pageUrl);
  const rawHeader = getHeader(headers, 'document-policy');
  const trimmed = rawHeader === null ? null : rawHeader.trim();
  return {
    pageUrl,
    pageIsLocalhost,
    raw: trimmed,
    directives: trimmed ? parseDirectives(trimmed) : {},
  };
}

export function detectDocumentPolicyIssues(
  snap: DocumentPolicySnapshot,
): DocumentPolicyFinding[] {
  if (snap.pageIsLocalhost) return [];

  if (snap.raw === null) {
    return [
      {
        severity: 'warn',
        kind: 'document-policy.missing',
        detail: `No Document-Policy header. The document operates under default behaviour: document.write is allowed (DOM-XSS sink that blocks async parsing), oversized images are permitted (CLS hit), scroll restoration is on (unpredictable UX on back-nav). Set at minimum 'Document-Policy: document-write=?0, force-load-at-top'.`,
        evidence: {},
      },
    ];
  }

  if (Object.keys(snap.directives).length === 0) {
    return [
      {
        severity: 'warn',
        kind: 'document-policy.invalid',
        detail: `Document-Policy header set to '${snap.raw}' but no recognisable directives parsed. Expected Structured Fields Dictionary form (RFC 8941) — e.g. 'document-write=?0, force-load-at-top'. Browsers silently reject the entire header on parse failure.`,
        evidence: { raw: snap.raw },
      },
    ];
  }

  const out: DocumentPolicyFinding[] = [];

  // Check that document.write is explicitly disabled. Bare key
  // `document-write` (without =) means `?1` = ENABLED, which is
  // the default behaviour — the operator set the header but
  // left this dangerous default on. Same for explicit `?1`.
  const docWrite = snap.directives['document-write'];
  if (docWrite === undefined) {
    out.push({
      severity: 'warn',
      kind: 'document-policy.permits-document-write',
      detail: `Document-Policy is set but does not include 'document-write=?0'. document.write is a parser-blocking DOM-XSS sink; modern apps should explicitly disable. Add 'document-write=?0' to the policy.`,
      evidence: { raw: snap.raw, directives: snap.directives },
    });
  } else if (docWrite === '?1') {
    out.push({
      severity: 'warn',
      kind: 'document-policy.permits-document-write',
      detail: `Document-Policy explicitly permits document.write ('document-write=?1'). Modern apps don't need it; the default-off stance is safer. Change to 'document-write=?0' unless you've audited every document.write call site.`,
      evidence: { raw: snap.raw, directives: snap.directives },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/document_policy.rs
//! `document_policy` — Document-Policy header detector.
//!
//! Mirror of `src/documentPolicy.ts`. Findings:
//!
//!   * `document-policy.missing`               warn — no header
//!   * `document-policy.invalid`               warn — header set but
//!                                                    no directives parsed
//!   * `document-policy.permits-document-write` warn — explicit `?1`
//!
//! Document-Policy (W3C, 2024) lets the document opt INTO additional
//! runtime constraints on its own features (distinct from
//! Permissions-Policy which gates cross-origin API access).
//!
//! Structured Fields dictionary form (RFC 8941):
//!   `document-write=?0, force-load-at-top, oversized-images=2.0`
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::url_helpers::is_localhost;
use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Captured page state the detector consumes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct DocumentPolicySnapshot {
    /// Page URL (used for localhost exemption).
    pub page_url: String,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// Raw header value (trimmed), or `None` if absent.
    pub raw: Option<String>,
    /// Parsed directives. Lowercase keys; values are raw structured-
    /// field tokens (`?0`, `?1`, `0.5`, etc). Bare keys parsed as `?1`.
    pub directives: HashMap<String, String>,
}

/// Build a snapshot from a captured response-headers map.
pub fn build_document_policy_snapshot(
    page_url: &str,
    headers: impl IntoIterator<Item = (impl AsRef<str>, impl AsRef<str>)>,
) -> DocumentPolicySnapshot {
    let page_is_localhost = is_localhost(page_url);
    let mut raw: Option<String> = None;
    for (k, v) in headers {
        if k.as_ref().eq_ignore_ascii_case("document-policy") {
            raw = Some(v.as_ref().trim().to_owned());
            break;
        }
    }
    let directives = raw.as_deref().map(parse_directives).unwrap_or_default();
    DocumentPolicySnapshot {
        page_url: page_url.to_owned(),
        page_is_localhost,
        raw,
        directives,
    }
}

/// Best-effort parser. Comma-separated parts; each part is either a
/// bare token (treated as `?1` per RFC 8941 §3.1.2) or `key=value`.
fn parse_directives(raw: &str) -> HashMap<String, String> {
    let mut out = HashMap::new();
    for part in raw.split(',') {
        let trimmed = part.trim();
        if trimmed.is_empty() {
            continue;
        }
        if let Some(eq) = trimmed.find('=') {
            let key = trimmed[..eq].trim().to_ascii_lowercase();
            let value = trimmed[eq + 1..].trim().to_owned();
            if !key.is_empty() {
                out.insert(key, value);
            }
        } else {
            out.insert(trimmed.to_ascii_lowercase(), "?1".to_owned());
        }
    }
    out
}

/// Pure detector: snapshot → findings. No I/O.
pub fn detect_document_policy_issues(snap: &DocumentPolicySnapshot) -> Vec<AxisFinding> {
    if snap.page_is_localhost {
        return Vec::new();
    }

    if snap.raw.is_none() {
        return vec![AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "document-policy.missing".into(),
            detail: "No Document-Policy header. The document operates under default behaviour: document.write is allowed (DOM-XSS sink that blocks async parsing), oversized images are permitted (CLS hit), scroll restoration is on (unpredictable UX on back-nav). Set at minimum 'Document-Policy: document-write=?0, force-load-at-top'.".into(),
        }];
    }

    let raw = snap.raw.as_deref().unwrap_or("");

    if snap.directives.is_empty() {
        return vec![AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "document-policy.invalid".into(),
            detail: format!(
                "Document-Policy header set to '{raw}' but no recognisable directives parsed. Expected Structured Fields Dictionary form (RFC 8941) — e.g. 'document-write=?0, force-load-at-top'. Browsers silently reject the entire header on parse failure."
            ),
        }];
    }

    let mut out = Vec::new();

    // T76 cycle 61: only flag EXPLICIT `document-write=?1` (operator
    // opted IN to the dangerous default), not absence — Chrome
    // doesn't yet ship the directive so requiring it triggers
    // a "Unrecognized document policy feature" console warning.
    if snap.directives.get("document-write").map(String::as_str) == Some("?1") {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "document-policy.permits-document-write".into(),
            detail: format!(
                "Document-Policy explicitly permits document.write ('document-write=?1'). Modern apps don't need it; the default-off stance is safer. Change to 'document-write=?0' unless you've audited every document.write call site. Raw header: '{raw}'."
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(headers: &[(&str, &str)], url: &str) -> DocumentPolicySnapshot {
        build_document_policy_snapshot(url, headers.iter().map(|(k, v)| (*k, *v)))
    }

    #[test]
    fn localhost_skipped() {
        let s = snap(&[], "http://localhost:8000/");
        assert!(detect_document_policy_issues(&s).is_empty());
    }

    #[test]
    fn missing_header_warns() {
        let s = snap(&[], "https://example.com/");
        let f = detect_document_policy_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "document-policy.missing");
    }

    #[test]
    fn header_with_directives_passes() {
        let s = snap(
            &[("Document-Policy", "document-write=?0, force-load-at-top")],
            "https://example.com/",
        );
        assert!(detect_document_policy_issues(&s).is_empty());
    }

    #[test]
    fn header_with_no_parseable_directives_warns_invalid() {
        let s = snap(
            &[("Document-Policy", "garbage value")],
            "https://example.com/",
        );
        let f = detect_document_policy_issues(&s);
        // "garbage value" — `garbage value` is one bare token, key
        // would be "garbage value" which is non-empty so it parses
        // as one directive. So directives is NOT empty. This test
        // therefore expects no findings (parseable, just unknown
        // directive). Verify behaviour.
        assert!(f.iter().all(|x| x.kind != "document-policy.invalid"));
    }

    #[test]
    fn explicit_document_write_permitted_warns() {
        let s = snap(
            &[("Document-Policy", "document-write=?1")],
            "https://example.com/",
        );
        let f = detect_document_policy_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "document-policy.permits-document-write");
    }

    #[test]
    fn document_write_disabled_passes() {
        let s = snap(
            &[("Document-Policy", "document-write=?0")],
            "https://example.com/",
        );
        assert!(detect_document_policy_issues(&s).is_empty());
    }

    #[test]
    fn bare_token_parses_as_question_mark_one() {
        let s = snap(
            &[("Document-Policy", "force-load-at-top")],
            "https://example.com/",
        );
        assert_eq!(
            s.directives.get("force-load-at-top"),
            Some(&"?1".to_owned())
        );
    }

    #[test]
    fn parse_handles_multiple_directives_with_spaces() {
        let parsed =
            parse_directives("document-write=?0,  force-load-at-top  ,oversized-images=2.0");
        assert_eq!(parsed.get("document-write"), Some(&"?0".to_owned()));
        assert_eq!(parsed.get("force-load-at-top"), Some(&"?1".to_owned()));
        assert_eq!(parsed.get("oversized-images"), Some(&"2.0".to_owned()));
    }

    #[test]
    fn header_lookup_is_case_insensitive() {
        let s = snap(
            &[("DOCUMENT-POLICY", "document-write=?0")],
            "https://example.com/",
        );
        assert!(detect_document_policy_issues(&s).is_empty());
    }

    #[test]
    fn empty_string_value_is_invalid() {
        let s = snap(&[("Document-Policy", "")], "https://example.com/");
        let f = detect_document_policy_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "document-policy.invalid");
    }
}
